client/client.js

```diff
@@ -62,71 +63,97 @@ Client.create = function (message) {
-      changed = false,
+Client.prototype.storeData = function (messageIn) {
+  var subCount = Object.keys(this.subscriptions).length,
+      presCount = Object.keys(this.presences).length,
+      logNow = false;
+
+  // Persist the message data, according to type
+  switch(messageIn.op) {
+    case 'unsubscribe':
+    case 'sync':
+    case 'subscribe':
+      logNow = this._storeDataSubscriptions(messageIn, subCount);
+      break;
+
+    case 'set':
+      logNow = this._storeDataPresences(messageIn);
+      break;
+  }
+
+  if (logNow) {
+    log.info('#storeData: client_id: ' + this.id +
+              '; subscription count: ' + subCount +
+              '; presence count: ' + presCount);
+  }
+
+  return true;
+};
+
+Client.prototype._storeDataSubscriptions = function (messageIn, subCount) {
+  var message = _cloneForStorage(messageIn),
+      to = message.to,
+      subscriptionHashname = this.key + '_subs',
+      logNow = false,
       lookupMessage;
 
   // Persist the message data, according to type
   switch(message.op) {
     case 'unsubscribe':
-      if (subscriptions[message.to]) {
-        delete subscriptions[message.to];
-        changed = true;
+      if (this.subscriptions[to]) {
+        delete this.subscriptions[to];
+        Core.Persistence.expire(subscriptionHashname, Client.getDataTTL());
+        Core.Persistence.deleteHash(subscriptionHashname, to);
+        logNow = 0 === subCount % LOG_WINDOW_SIZE;
       }
       break;
 
     case 'sync':
     case 'subscribe':
-      lookupMessage = subscriptions[message.to];
+      lookupMessage = this.subscriptions[to];
       if (!lookupMessage ||
             (lookupMessage.op != 'sync' && !_.isEqual(lookupMessage, message))) {
-        subscriptions[message.to] = message;
-        changed = true;
+        this.subscriptions[to] = message;
+        Core.Persistence.expire(subscriptionHashname, Client.getDataTTL());
+        Core.Persistence.persistHash(subscriptionHashname, to, message);
+        logNow = 0 === subCount % LOG_WINDOW_SIZE;
       }
       break;
+  }
+
+  return logNow;
+};
+
+Client.prototype._storeDataPresences = function (messageIn, presCount) {
+  var message = _cloneForStorage(messageIn),
+      to = message.to,
+      presenceHashname = this.key + '_pres',
+      logNow = false,
+      lookupMessage;
 
+  // Persist the message data, according to type
+  switch(message.op) {
     case 'set':
-      if (message.to.substr(0, 'presence:/'.length) == 'presence:/') {
-        lookupMessage = presences[message.to];
+      if (to.substr(0, 'presence:/'.length) == 'presence:/') {
+        lookupMessage = this.presences[to];
         if (!lookupMessage || (!_.isEqual(lookupMessage, message))) {
-          presences[message.to] = message;
-          changed = true;
+          this.presences[to] = message;
+          Core.Persistence.expire(presenceHashname, Client.getDataTTL());
+          Core.Persistence.persistHash(presenceHashname, to, message);
```

Comment:
This is a problem. We will end up storing offlines here, and the hash will continue to bloat with nothing removing it. I suggest we remove offline presences.